# Probability Basics

## Experiment and outcomes

Previously, we defined sets and related concepts. Now let's look at the set

$S= \{1,2,3,4,5,6\}$, which contains all possible outcomes when throwing a dice.

When you throw a dice once, you can consider this a *random experiment*. The result of this "experiment" is the *outcome*. 

You can then say that:

- $S$ defines all the **possible outcomes** when throwing the dice once
- $S$ is our Universal set $\Omega$, as seen before


When conducting experiments, you say that your universal set is your **sample space**: it is the universe in which your possible outcomes are listed as elements.

Other examples of sample spaces:
- The number of text messages you send each day:  in this case, S is equal to some number x, with x being a **positive integer**, or mathematically: $S = \{x \mid x \in \mathbb{Z}, x \geq 0\}$
- The number of hours someone watches TV each day:  $S = \{x \mid x \in \mathbb{R}, 0 \leq x \leq 24 \}$

## Event space

Next, let's define event space. The **event space** is a subset of the sample space, $E\subseteq S$

For example, the event "throwing a number higher than 4" when throwing a dice would result in an event space $E= \{5,6\}$. Throwing an odd number would lead to an event space $E= \{1,3,5\}$. 

Summarized, the event space is a collection of events that we *care* about. We say that event $E$ happened if the actual outcome after rolling the dice belongs to the predefined event space  $E$.

With **sample space** and **event space**, you now understand the two foundational concepts of **probability**. 

Other examples of event spaces based on previously defined sample spaces:

- If you define that the event "low daily number of text messages sent" means 20 or fewer text messages, the event space is defined as: $E = \{x \mid x \in \mathbb{Z}, 0 \leq x \leq 20 \}$
- Binge-watch day:  $E = \{x \mid x \in \mathbb{R}, x \geq 6 \}$

## Introduction to probability

### The law of relative frequency

While conducting an endless stream of experiments, the relative frequency by which an event will happen becomes a fixed number. 

Let's denote an event by $E$, and the _probability_ of the event $E$ occurring by $P(E)$. Next, let $n$ be the number of conducted experiments, and $S(n)$ the count of "successful" experiments (i.e. the times that event $E$ happened). The formal definition of probability as a relative frequency is given by:

$$P(E) = \lim_{n\rightarrow\infty} \dfrac{S{(n)}}{n}$$

This is the basis of a frequentist statistical interpretation: an event's probability is the ratio of the positive trials to the total number of trials as we repeat the process infinitely. 

For example, the probability of rolling a 5 on a 6 sided dice is the limit of the successes to trials as the number of trials goes to infinity.


###  Probability axioms

In the early 20th century, Kolmogorov and Von Mises came up with three axioms that further expand on the idea of probability. The three axioms are:

#### 1. Positivity

A probability is always bigger than or equal to 0, or $0 \leq P(E) \leq 1$

#### 2. Probability of a certain event

If the event of interest is the sample space, we say that the outcome is a certain event, or $P(S) = 1$

#### 3. Additivity 

The probability of the union of two exclusive events is equal to the sum of the probabilities of the individual events happening.

If $A \cap B = \emptyset $, then $P(A\cup B) = P(A) + P(B)$ 

### Addition law of probability

The additivity axiom is great, but most of the time events are not exclusive. A very important property is the **addition law or probability** or the **sum rule**.

$P(A\cup B) = P(A) + P(B) - P(A \cap B) $ 

Put in words, the probability that $A$ or $B$ will happen is the sum of the probabilities that $A$ will happen and that $B$ will happen, minus the probability that *both* $A$ and $B$ will happen.

## Examples

Let's reconsider the dice example to explain what was explained before:

### Additivity of exclusive events

Let's consider two events: event $M$ means throwing a 6, event $N$ means that you throw an odd number $N={1,3,5}$. These events are exclusive, and you can use the additivity rule if you want to know the answer to the question: 

*"what is the probability that your outcome will be a 6, or an odd number?"*

$P(M\cup N) = P(M) + P(N) = \dfrac{1}{6}+\dfrac{3}{6}=\dfrac{4}{6} $ 

### Addition law of probability

Now, let's consider the same event $N={1,3,5}$ and another event $Q={4,5}$. These events are *not* mutually exclusive, so if you want to know the probability that $N$ or $Q$ will happen, you need to use the addition law of probability.

Note that $(N \cap Q)$ is equal to getting an outcome of 5, as that is the "common" element in the respective event spaces of $N$ and $Q$. This means that $P(N \cap Q) = \dfrac{1}{6}$

$P(N\cup Q) = P(N) + P(Q) - P(N \cap Q) = \dfrac{3}{6} + \dfrac{2}{6} - \dfrac{1}{6} = \dfrac{4}{6} $ 


## Final Note

In the previous examples, you noticed that for our dice example, it is easy to use these fairly straightforward probability formulas to calculate probabilities of certain outcomes. 

However, if you think about our text message example, things are less straightforward, e.g.:

*"What is the probability of sending less than 20 text messages in a day?"*

This is where the probability concepts introduced here fall short. The probability of throwing any number between 1 and 6 with a die is always exactly $\dfrac{1}{6}$,  but we can't simply count our messages event space. In words, the probability of sending 20 messages is likely different than the probability of sending, say, 5 messages, and will be different for any number of messages sent. You'll learn about tools to solve problems like these later on.

---

## Practice

##  Sample space, event space and the law of relative frequency

#### a. Let's throw a dice once

First, create a set `roll_dice` that holds the sample space of rolling a 6-sided dice once.

In [ ]:
roll_dice = set([1,2,3,4,5,6])

Now, let's assume that the event space is defined by "throwing a number higher than 4". This means that we consider the outcome "successful" if a 5 or a 6 is thrown. Create a set that holds these values.

In [ ]:
event = set([5,6])

Now use the formula $P(E) = \dfrac{|E|}{|S|}$ (This formula is called "Laplace's formula" and strongly related to the law of relative frequency) to calculate the probability.

In [ ]:
prob_5_6 = len(event)/len(roll_dice)
prob_5_6  # 0.3333333333333333

0.3333333333333333

Using this formula, it should be clear that the answer is 1/3 or 0.3333....  

#### b. Now, let's simulate rolling dice to see how the law of relative frequency works.

The law of relative frequency can be used to prove certain probabilities. But how does this work exactly? You're about to find out!

$$P(E) = \lim_{n\rightarrow\infty} \dfrac{S{(n)}}{n}$$

As you can see in the formula, the law states that when repeating an experiment $n$ times, where $n$ is very big, and you divide the number of "good" outcomes by the sample space (here we call it event E), you get to the probability of the event E. It should be clear that we get a more accurate number for P(E) when $n$ grows.

Let's see how this works. First, let's randomly generate values between 1 and 6. You can use `numpy` (imported as `np`) to generate random integers between 1 and 6 by setting the correct arguments. The `np.random` module is a very useful tool for this. We helped you with the code here, but you'll get more practice and a thorough explanation later on!

In [ ]:
import numpy as np
np.random.randint(1,7) # you will get a random value between 1 and 6. See how it changes when you rerun

3

Now, let's repeat this experiment 10 times, then 1000 times, then 1 million times, then 100 million times. 
You can do this by specifying the argument `size` within the numpy function used above. Store the values in the pre-defined variables below.

In [ ]:
np.random.seed(12345) # to ensure reproducibility of results

dice_10 = np.random.randint(1,7,size=10)
dice_1k = np.random.randint(1,7,size=1000)
dice_1m = np.random.randint(1,7,size=1000000)
dice_100m = np.random.randint(1,7,size=100000000)

In [ ]:
dice_10

array([3, 6, 6, 2, 5, 2, 6, 3, 6, 2])

Next, let's count the number of "events". Remember that an event here is defined as throwing a 5 or a 6. Store them in the values below.

In [ ]:
event_10 = np.sum(dice_10>4)
event_1k = np.sum(dice_1k>4)
event_1m = np.sum(dice_1m>4)
event_100m = np.sum(dice_100m>4)

Next, you'll divide the number of events for each $n$ by the respective values for $n$. What do you see?

In [ ]:
prob_10 = event_10/10
prob_1k = event_1k/1000
prob_1m = event_1m/1000000
prob_100m = event_100m/100000000
prob_10, prob_1k, prob_1m, prob_100m # 0.5 0.331 0.333657 0.33329752

(0.5, 0.331, 0.333657, 0.33329752)

You see that the probability converges to 0.3333333... for higher values of $n$. 

##  The Probability Axioms

You're working at the United Nations, and want to get a better sense of the world population. 

You come across some numbers and find the list of probabilities of being an inhabitant for each of the seven continents (rounded up to 3 digits):

- P(Africa) = 0.161
- P(Antarctica) = 0.000
- P(Asia) = 0.598
- P(Europe) = 0.10
- P(North-America) = 0.079
- P(Australia) = 0.005
- P(South-America) = 0.057

store these values using the variable names below:

In [ ]:
P_afr = 0.161
P_ant = 0.000
P_as = 0.598
P_eur = 0.1
P_na = 0.079
P_aus = 0.005
P_sa = 0.057

Now create the sample space set names `continents`. Store the sample space in a numpy array.

In [ ]:
continents = np.array([P_afr, P_ant, P_as, P_eur, P_na, P_aus, P_sa])
print(continents)

[0.161 0.    0.598 0.1   0.079 0.005 0.057]


We want to make sure that the three probability axioms are fulfilled because they assure us that $(\Omega,E,P)$ is a **probability space**:

- if we have a sample space $S$ (or $\Omega$)
- if we have an event space $E$ and a probability measure $P$, 
- **and** the three probability axioms introduced in the previous lesson are fulfilled, 

The third axiom is fairly ad hoc, and you will basically have to deduce from the context whether individual events are independent. It is fairly straightforward, however, that people cannot be inhabitants of two continents at the same time, so for now, we will assume that we're good for axiom three.

However, we can use the numpy array `continents` to verify if axiom 1 and 2 are fulfilled. Create a function `check_axioms` with a single input, `sample_space`, that returns the message "We're good!" if both axiom 1 and 2 are fulfilled, and "Not quite!" if that's not the case.

In [ ]:
def check_axioms(sample_space):
    test_ax1a = (sample_space >= 0).all()
    test_ax1b = (sample_space <= 1 ).all()
    test_ax2 = np.sum(sample_space) == 1
    if (test_ax1a == True) & (test_ax1b == True) & (test_ax2 == True):
        return("We're good!")
    else:
        return("Not quite!")

Now test your newly created function out on the sample space `continents`:

In [ ]:
check_axioms(continents)

"We're good!"

You want to make sure your test returns `"Not quite!"` for the following numpy arrays. Go ahead and test away!

In [ ]:
test_1 = np.array([0.05, 0.2, 0.3, 1.01])
test_2 = np.array([0.05, 0.5, 0.6, -0.15])
test_3 = np.array([0.043,0.05,.02,0.3,0.2])

In [ ]:
print(check_axioms(test_1))
print(check_axioms(test_2))
print(check_axioms(test_3))

Not quite!
Not quite!
Not quite!


Great! We tested it and seems like our set `continents` is a true probability space.

## Some more practice on the sample and event spaces

In this exercise, we'll look at possible outcomes when throwing a dice twice. For your convenience, we created the NumPy array below.

Next, we'll compute a couple or probabilities associated with doing this.

In [ ]:
import numpy as np
sample_dice = np.array([(1, 1), (1, 2), (1, 3), (1, 4), (1, 5), (1, 6),
              (2, 1), (2, 2), (2, 3), (2, 4), (2, 5), (2, 6),
              (3, 1), (3, 2), (3, 3), (3, 4), (3, 5), (3, 6),
              (4, 1), (4, 2), (4, 3), (4, 4), (4, 5), (4, 6),
              (5, 1), (5, 2), (5, 3), (5, 4), (5, 5), (5, 6),
              (6, 1), (6, 2), (6, 3), (6, 4), (6, 5), (6, 6)])

Look at the shape of the array to reassure we haven't made any mistakes.

In [ ]:
sample_dice.shape  # should be equal to (36,2)

(36, 2)

Use Python to obtain the following probabilities:

#### a. What is the probability of throwing a 5 at least once?

First, use `sample_dice` to get "True" values for each time a 5 occurs.

In [ ]:
set_5 = sample_dice == (5) # Your output should be an array of shape (36, 2) with booleans instead of numbers
set_5.shape

(36, 2)

Next, make sure that you get a value `True` for each pair where at least one 5 was thrown.

In [ ]:
true_5 = np.any(set_5, axis = 1) # Your output should be an array of shape (36,) and have booleans. 
              # You should obtain "True" if at least one of the previous pairs was true. 
              # Hint: Use np.any()
print(true_5)

[False False False False  True False False False False False  True False
 False False False False  True False False False False False  True False
  True  True  True  True  True  True False False False False  True False]


Applying the `sum()` function to this result you can get to the total number of items in the event space. Divide this number by the total number in the sample space to obtain the probability of throwing a 5 at least once.

In [ ]:
prob_5 = None
print(prob_5)

#### b. What is the probability of throwing a 5 or 6 at least once?

In [ ]:
set_5 =  sample_dice == (5)
set_6 = sample_dice == (6)

In [ ]:
set_5_6 = (set_6 + set_5)

In [ ]:
set_any_5_6 = set_5_6.any(axis=1)
print(set_any_5_6) 

[False False False False  True  True False False False False  True  True
 False False False False  True  True False False False False  True  True
  True  True  True  True  True  True  True  True  True  True  True  True]


In [ ]:
prob_5_6 = set_any_5_6.sum()/len(sample_dice)
print(prob_5_6)

0.5555555555555556


#### c. What is the probability of the outcome having a sum of exactly 8?

In [ ]:
sum_dice = np.sum(sample_dice, axis = 1)
sum_8 = sum(sum_dice == 8)

In [ ]:
prob_sum_8 = sum_8/len(sample_dice)
print(prob_sum_8)

0.1388888888888889


## Now let's try creating your own event space!

A teaching assistant is holding office hours so students can make appointments. She has 6 appointments scheduled today, 3 by male students, and 3 by female students. 

Create a NumPy array of all possible orders of three male and three female students in which the teaching assistant can see students by appointment (please note: only consider gender when creating event space). Create this NumPy array in the same way as we did in the "throwing a dice twice" exercise. It will be quite a bit of typing, as your resulting NumPy array will have a shape (20,6)!

In [ ]:
sample_mf= np.array([("M","M","M","F","F","F"), ("M","M","F","M","F","F"), ("M","M","F","F","M","F"),
                     ("M","M","F","F","F","M"), ("M","F","M","M","F","F"), ("M","F","M","F","F","M"),
                     ("M","F","M","F","M","F"), ("M","F","F","M","F","M"), ("M","F","F","M","M","F"),
                     ("M","F","F","F","M","M"), ("F","F","F","M","M","M"), ("F","F","M","F","M","M"), 
                     ("F","F","M","M","F","M"), ("F","F","M","M","M","F"), ("F","M","F","F","M","M"),
                     ("F","M","F","M","M","F"), ("F","M","F","M","F","M"), ("F","M","M","F","M","F"),
                     ("F","M","M","F","F","M"), ("F","M","M","M","F","F") ])

In [ ]:
sample_mf.shape # get the shape of sample_mf

(20, 6)

In [ ]:
sample_length= len(sample_mf)
print(sample_length)

20


#### 1. Calculate the probability that at least 2 out of the first 3 appointments are with female students

First, select the first 3 appointment slots and check for "F".

In [ ]:
first_3_F = sample_mf[:,:3] == "F"
print(first_3_F)

[[False False False]
 [False False  True]
 [False False  True]
 [False False  True]
 [False  True False]
 [False  True False]
 [False  True False]
 [False  True  True]
 [False  True  True]
 [False  True  True]
 [ True  True  True]
 [ True  True False]
 [ True  True False]
 [ True  True False]
 [ True False  True]
 [ True False  True]
 [ True False  True]
 [ True False False]
 [ True False False]
 [ True False False]]


In [ ]:
num_F = np.sum(first_3_F, axis=1)
print(num_F)

[0 1 1 1 1 1 1 2 2 2 3 2 2 2 2 2 2 1 1 1]


In [ ]:
F_2plus = np.sum(num_F > 1)
print(F_2plus)

10


In [ ]:
prob_F_2plus = F_2plus.sum()/sample_length
print(prob_F_2plus)

0.5


#### 2. Calculate the probability that after 4 appointment slots, all the female students have had an appointment

In [ ]:
np.sum((sample_mf[:,4:]== ['M','M']).all(axis=1))/sample_length

0.2

You noticed that coming up with the sample space was probably the most time-consuming part of the exercise, and it would really become unfeasible to write this down for say, 10 or, even worse, 20 appointments in a row. You'll learn about methods that make this process easier soon!

## The Addition Law of Probability
At a supermarket, we randomly select customers, and make notes of whether a certain customer owns a Visa card (event A) or an American Express credit card (event B). Some customers own both cards.
You can assume that:

- P(A) = 0.5
- P(B) = 0.4
- both A and B = 0.25.

1) Compute the probability that a selected customer has at least one credit card.

2) Compute the probability that a selected customer doesn't own any of the mentioned credit cards.

3) Compute the probability that a customer *only* owns VISA card.

(You can use Python here, but you don't have to)

In [ ]:
# 1) $P(A\cup B) = P(A) + P(B) - P(A\cap B)  = 0.5+0.4-0.25 = 0.65 $

# 2) $1 - P(A\cup B) = 0.35 $

# 3) $P(A)-P(B\cap A) = 0.25$